File: packages/langchain/src/message-history.ts
// ─────────────────────────────────────────────
// MemHQMessageHistory
//
// Implements LangChain.js BaseListChatMessageHistory so any chain or
// agent that takes a chat history (RunnableWithMessageHistory,
// ConversationChain, AgentExecutor) can read/write directly from MemHQ.
// ─────────────────────────────────────────────

import { BaseListChatMessageHistory } from "@langchain/core/chat_history";
import {
  AIMessage,
  HumanMessage,
  SystemMessage,
  ToolMessage,
  type BaseMessage,
} from "@langchain/core/messages";
import { MemoryClient, type Role } from "@memhq/sdk";

export interface MemHQMessageHistoryOptions {
  /** MemHQ API key. Falls back to MEMHQ_API_KEY env var. */
  apiKey?: string;
  /** Session identifier — typically the LangChain session_id. Stored as MemHQ thread metadata. */
  sessionId: string;
  /** External user id. Required — MemHQ is user-scoped. */
  userId: string;
  /** Override the MemHQ base URL (default https://api.memhq.ai). */
  baseUrl?: string;
  /** How many top memories to prepend on `getMessages()`. Default 10. */
  recallLimit?: number;
  /** Reuse an existing MemoryClient (skips constructing a new one). */
  client?: MemoryClient;
}

/**
 * Chat message history backed by MemHQ. Reads return recent memories
 * formatted as messages; writes ingest into the user's graph.
 *
 * The "history" here is semantic, not chronological — `getMessages()`
 * surfaces the memories most relevant to the latest exchange, not raw
 * transcript. This is the right shape for agents whose context window
 * can't hold the full conversation.
 */
export class MemHQMessageHistory extends BaseListChatMessageHistory {
  lc_namespace = ["memhq", "chat_history"];

  private readonly client: MemoryClient;
  private readonly sessionId: string;
  private readonly userId: string;
  private readonly recallLimit: number;

  constructor(opts: MemHQMessageHistoryOptions) {
    super();
    this.sessionId = opts.sessionId;
    this.userId = opts.userId;
    this.recallLimit = opts.recallLimit ?? 10;
    this.client =
      opts.client ??
      new MemoryClient({
        ...(opts.apiKey !== undefined ? { apiKey: opts.apiKey } : {}),
        ...(opts.baseUrl !== undefined ? { baseUrl: opts.baseUrl } : {}),
      });
  }

  /**
   * Returns the top-k semantically relevant memories for this user,
   * shaped as LangChain messages. Used by RunnableWithMessageHistory
   * to prepend to the model prompt.
   */
  async getMessages(): Promise<BaseMessage[]> {
    // Use the session id itself as a recency anchor — pulls memories
    // most associated with this thread. Callers wanting more targeted
    // retrieval should use MemHQRetriever directly.
    const search = await this.client.search({
      query: this.sessionId,
      userId: this.userId,
      limit: this.recallLimit,
    });
    return search.results.map((m) => new SystemMessage({ content: m.content }));
  }

  /** Append a single message to MemHQ. */
  async addMessage(message: BaseMessage): Promise<void> {
    const role = mapRole(message);
    const content = stringifyContent(message.content);
    await this.client.add({
      userId: this.userId,
      messages: [{ role, content }],
      metadata: { session_id: this.sessionId },
    });
  }

  /** Batch ingest — preferred for whole-turn writes. */
  override async addMessages(messages: BaseMessage[]): Promise<void> {
    if (messages.length === 0) return;
    await this.client.add({
      userId: this.userId,
      messages: messages.map((m) => ({
        role: mapRole(m),
        content: stringifyContent(m.content),
      })),
      metadata: { session_id: this.sessionId },
    });
  }

  /**
   * MemHQ does not expose a session-scoped delete via the hosted API
   * yet. This is a no-op so callers don't blow up; the next add() will
   * append normally. To wipe a user wholesale, use
   * `client.users.delete(userId)` from `@memhq/sdk`.
   */
  override async clear(): Promise<void> {
    // intentional no-op
  }
}

function mapRole(message: BaseMessage): Role {
  if (message instanceof HumanMessage) return "user";
  if (message instanceof AIMessage) return "assistant";
  if (message instanceof SystemMessage) return "system";
  if (message instanceof ToolMessage) return "tool";
  // Fallback for custom message classes — treat as user input so the
  // extractor still indexes the content.
  return "user";
}

function stringifyContent(content: BaseMessage["content"]): string {
  if (typeof content === "string") return content;
  // Multi-part content (vision, tool calls) → flatten the text parts.
  return content
    .map((part) => {
      if (typeof part === "string") return part;
      if (part.type === "text") return (part as { text: string }).text;
      return "";
    })
    .filter(Boolean)
    .join("\n");
}

File: python/memhq-langchain/memhq_langchain/message_history.py
"""MemHQ-backed chat message history.

Implements ``langchain_core.chat_history.BaseChatMessageHistory`` so the
adapter slots into ``RunnableWithMessageHistory`` and any chain that
expects a chat history object. Reads return the top-k memories most
relevant to the session, shaped as LangChain messages.
"""

from __future__ import annotations

from typing import List, Optional

from langchain_core.chat_history import BaseChatMessageHistory
from langchain_core.messages import (
    AIMessage,
    BaseMessage,
    HumanMessage,
    SystemMessage,
    ToolMessage,
)

from memhq_langchain._client import _Transport, add_messages, search


def _role(message: BaseMessage) -> str:
    if isinstance(message, HumanMessage):
        return "user"
    if isinstance(message, AIMessage):
        return "assistant"
    if isinstance(message, SystemMessage):
        return "system"
    if isinstance(message, ToolMessage):
        return "tool"
    return "user"


def _stringify(content: object) -> str:
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        parts = []
        for part in content:
            if isinstance(part, str):
                parts.append(part)
            elif isinstance(part, dict) and part.get("type") == "text":
                parts.append(str(part.get("text", "")))
        return "\n".join(p for p in parts if p)
    return str(content)


class MemHQChatMessageHistory(BaseChatMessageHistory):
    """LangChain chat history backed by MemHQ.

    Parameters
    ----------
    api_key:
        MemHQ API key. Falls back to ``MEMHQ_API_KEY``.
    session_id:
        Session identifier used as the MemHQ thread anchor.
    user_id:
        External user id. Required — MemHQ is user-scoped.
    base_url:
        Override the API base URL.
    recall_limit:
        How many memories to surface on each history read. Default 10.
    """

    def __init__(
        self,
        *,
        session_id: str,
        user_id: str,
        api_key: Optional[str] = None,
        base_url: Optional[str] = None,
        recall_limit: int = 10,
    ) -> None:
        self._transport = _Transport(api_key=api_key, base_url=base_url)
        self._session_id = session_id
        self._user_id = user_id
        self._recall_limit = recall_limit

    @property
    def messages(self) -> List[BaseMessage]:  # type: ignore[override]
        """Return relevant memories shaped as messages.

        Note: semantic, not chronological. The MemHQ wedge is that
        recall is reranked by relevance — agents don't need to hold
        the full transcript.
        """
        results = search(
            self._transport,
            query=self._session_id,
            user_id=self._user_id,
            limit=self._recall_limit,
        )
        return [SystemMessage(content=r.get("content", "")) for r in results]

    def add_message(self, message: BaseMessage) -> None:
        add_messages(
            self._transport,
            user_id=self._user_id,
            messages=[{"role": _role(message), "content": _stringify(message.content)}],
            metadata={"session_id": self._session_id},
        )

    def add_messages(self, messages: List[BaseMessage]) -> None:  # type: ignore[override]
        if not messages:
            return
        add_messages(
            self._transport,
            user_id=self._user_id,
            messages=[
                {"role": _role(m), "content": _stringify(m.content)} for m in messages
            ],
            metadata={"session_id": self._session_id},
        )

    def clear(self) -> None:
        """No-op. MemHQ does not expose session-scoped delete via the
        hosted API yet. To wipe a user wholesale, use the official
        ``memhq`` SDK's ``users.delete`` method.
        """
        return None
